File: 10699 Count the factors/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#define MAX_NUM 1000
#define GET(n) (mark[(n) >> 5] & (1u << ((n) & 31)))
#define SET(n) (mark[(n) >> 5] |= (1u << ((n) & 31)))

unsigned int mark[(MAX_NUM >> 6) + 1];
short *primes;
int size = 1;

void eratosthenesSieve() {
    primes = (short*)malloc(sizeof(short));
    primes[0] = 2;
    for(short n = 3; n <= MAX_NUM; n += 2) {
        if(!GET(n >> 1)) {
            primes = (short*)realloc(primes, (size + 1) * sizeof(short));
            primes[size++] = n;
            if(n <= 31) {
                for(int i = n * n; i <= MAX_NUM; i += n << 1)
                    SET(i >> 1);
            }
        }
    }
}

int main() {
    eratosthenesSieve();
    int num;
    while(scanf("%d", &num) && num != 0) {
        printf("%d : ", num);
        short limit = sqrt(num);
        int count = 0;
        for(int i = 0; i < size && primes[i] <= limit; ++i) {
            if(num % primes[i] == 0) {
                ++count;
                do {
                    num /= primes[i];
                } while(num % primes[i] == 0);
                limit = sqrt(num);
            }
        }
        printf("%d\n", count + (num > 1));
    }
    free(primes);
    return 0;
}

File: 10699 Count the factors/solution.java
import java.io.*;
import java.util.*;

public class Main {
    static short MAX_NUM = 1000;
    static boolean[] isComposite = new boolean[(MAX_NUM >> 1) + 1];
    static ArrayList<Short> primes = new ArrayList<>();
	public static void main (String[] args) throws IOException {
		eratosthenesSieve();
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
		int num;
		StringBuilder output = new StringBuilder();
		while(st.nextToken() == StreamTokenizer.TT_NUMBER && (num = (int)st.nval) != 0) {
		    output.append(num).append(" : ");
		    short limit = (short)Math.sqrt(num);
		    int count = 0;
		    for(int i = 0; i < primes.size() && primes.get(i) <= limit; ++i) {
		        if(num % primes.get(i) == 0) {
                    ++count;
                    do {
                        num /= primes.get(i);
                    } while(num % primes.get(i) == 0);
                    limit = (short)Math.sqrt(num);
		        }
		    }
		    output.append(count + (num > 1 ? 1 : 0)).append('\n');
		}
		System.out.print(output);
	}
	
	static void eratosthenesSieve() {
	    primes.add((short)2);
        for(short n = 3; n <= MAX_NUM; n += 2) {
            if(!isComposite[n >> 1]) {
                primes.add(n);
                if(n <= 31) {
                    for(int i = n * n; i <= MAX_NUM; i += n << 1)
                        isComposite[i >> 1] = true;
                }
            }
        }
    }
}
